### ISSUE TITLE
Multiple issues with big queries

### ISSUE THREAD

**Issue by giner** [OpenTSDB/opentsdb#668]
Hi,

There is a long time running instance of OpenTSDB 2.1.0-RC2 with default configuration. If somebody runs a big query, for example, to get some data for last year, then:

1) web interface fails quite quick (in about a minute) with "Request failed"
2) while web interface fails query itself continues which is visible by monitoring system load (I/O and CPU) and it can continue for very long time
3) until this query stops TSDB does not process any data from telnet interface which results in TCP connections to telnet interface getting stuck when socket buffer is full

The biggest problem is point 3 and it seems as a bug. At the same time being able to stop a long query forcefully would be nice too.

Best regards,
Stanislav

**Comment by giner** [comment]
Update: problem mentioned under number 3 is surprisingly solved by disabling network async I/O:
tsd.network.async_io = false

**Comment by giner** [comment]
Update: problem mentioned under number 1 is irrelevant and related to infrastructure itself. Just the same as #407 

The only problem left is under number 2 - when browser connection lost it make no sense to continue gathering data and tsdb should recognize and stop it.

**Comment by giner** [comment]
> The only problem left is under number 2 - when browser connection lost it make no sense to continue gathering data and tsdb should recognize and stop it.

I'm talking to myself :). But anyway, I've just realized that this behavior actually can be useful when you do queries and your connection times out for some reason (for me it is being behind load balancer with 60 seconds timeout). If session times out I can try again in 10 minutes and get the result from cache.

But some kind of keepalive would be better so that load balancer would not kill the connection. Possible?

**Comment by manolama** [comment]
Yeah we definitely should kill the query when the socket is closed, that's something I may work on soon as we see similar problems though, yeah it can be a little useful to reconnect and read the results from cache. So we could make it a configurable behavior. Ultimately we should stream data for large queries.

**Comment by johann8384** [comment]
Streaming data for large queries is sometimes problematic too just because of the aggregation. One thing I noticed a while back is that long running queries were due to an OpenTSDB instance on Node A and the region being on Node B. We had large datasets that were ultimately aggregated out but still had to be read from the region. So we'd transfer 1200GB of data (9600Gb / 10Gbps / 60 seconds =  16 minutes) from the regionserver to OpenTSDB and then aggregate it all away and return 2MB to the client. This resulted in a 16m20s query time. To resolve it we wrote a little proxy in the front that was a member of the HBase cluster and would know what region a metric was stored on so it could send the request to the OpenTSDB instance on that specific region. I *think* it did a 302 redirect to tell the client to ask the right OpenTSDB instance but it's been 8 years now so I am unsure. (https://github.com/turn/splicer)
